lib/ASTGen/Sources/ASTGen/Attrs.swift

```diff
@@ -0,0 +1,1132 @@
+    guard
+      var args = node.arguments?.as(LabeledExprListSyntax.self)?[...]
+    else {
+      // TODO: Diagnose.
+      return nil
+    }
+    fatalError("unimplemented")
+  }
+
+  func generateStringLiteralTextIfNotInterpolated(expr node: some ExprSyntaxProtocol) -> BridgedStringRef? {
+    if let segments = node.as(SimpleStringLiteralExprSyntax.self)?.segments {
+      return extractRawText(segments).bridged
+    } else if let segments = node.as(StringLiteralExprSyntax.self)?.segments,
+      segments.all({ $0.is(StringSegmentSyntax.self) })
+    {
+      return extractRawText(segments).bridged
+    }
+    return nil
+  }
+
+  func generateConsumingAttrOption<R>(
+    args: inout Slice<LabeledExprListSyntax>,
+    label: SyntaxText?,
+    _ valueGeneratorFunction: (ExprSyntax) -> R?
+  ) -> R? {
+    guard let arg = args.first else {
+      // TODO: Diagnose.
+      return nil
+    }
+    guard arg.label?.rawText == label else {
+      // TODO: Diagnose.
+      return nil
+    }
+    // Label matched. Consume the argument even if the value is not valid.
+    args.removeFirst()
+
+    return valueGeneratorFunction(arg.expression)
+  }
+
+  func generateConsumingPlainIdentifierAttrOption<R>(
+    args: inout Slice<LabeledExprListSyntax>,
+    _ valueGeneratorFunction: (TokenSyntax) -> R?
+  ) -> R? {
+    return generateConsumingAttrOption(args: &args, label: nil) {
+      guard let declRefExpr = $0.as(DeclReferenceExprSyntax.self),
+            declRefExpr.argumentNames == nil else {
+        // TODO: Diagnose.
+        return nil
+      }
+      return valueGeneratorFunction(declRefExpr.baseName)
+    }
+  }
+
+  func generateConsumingSimpleStringLiteralAttrOption(
+    args: inout Slice<LabeledExprListSyntax>,
+    label: SyntaxText? = nil
+  ) -> BridgedStringRef? {
+    return self.generateConsumingAttrOption(args: &args, label: label) {
+      self.generateStringLiteralTextIfNotInterpolated(expr: $0)
+    }
+  }
+
+  /// Extracts single identifier token argument from the `attribute`, calls the
+  /// specified function with the token, then returns the result. If `valueIfOmitted`
+  /// is not nil, and there's no argument clause in the `attribute`, this returns
+  /// the `valueIfOmitted` value.
+  /// Returns `nil` if the argument was not a single token, the callback returned `nil`,
+  /// or there was no argument clause and `valueIfOmitted` is not specified. In such cases,
+  /// this emits diagnostics.
+  func generateSingleAttrOption<Result>(
+    attribute node: AttributeSyntax,
+    _ valueGeneratorFunction: (TokenSyntax) -> Result?,
+    valueIfOmitted: Result? = nil
+  ) -> Result? {
+    guard node.leftParen != nil, let arguments = node.arguments else {
+      if let valueIfOmitted {
+        return valueIfOmitted
+      }
+      // TODO: Diagnose.
+      return nil
+    }
+
+    guard var arguments = arguments.as(LabeledExprListSyntax.self)?[...] else {
+      // TODO: Diagnose.
+      return nil
+    }
```

Comment:
Would it make sense to have some function that does the casting to `LabeledExprListSyntax` invokes a closure to run all the `generateConsuming` functions and the diagnose `if !arguments.isEmpty`? It seems very easy to forget the `if !arguments.isEmpty` check.